core/internal/clients/retry_policy.go

```diff
@@ -15,16 +14,18 @@ const CtxRetryPolicyKey ContextKey = "retryFunc"
 func DefaultRetryPolicy(ctx context.Context, resp *http.Response, err error) (bool, error) {
 	statusCode := resp.StatusCode
 	switch {
-	case statusCode == http.StatusBadRequest, statusCode == http.StatusConflict: // don't retry on 400 bad request or 409 conflict
-		return false, fmt.Errorf("the server responded with an error. (Error %d: %s)", statusCode, http.StatusText(statusCode))
+	case statusCode == http.StatusBadRequest: // don't retry on 400 bad request or 409 conflict
+		return false, err
+	case statusCode == http.StatusConflict:
+		return false, err
 	case statusCode == http.StatusUnauthorized: // don't retry on 401 unauthorized
-		return false, fmt.Errorf("the API key you provided is either invalid or missing. (Error %d: %s)", statusCode, http.StatusText(statusCode))
+		return false, err
 	case statusCode == http.StatusForbidden: // don't retry on 403 forbidden
-		return false, fmt.Errorf("you don't have permission to access this resource. (Error %d: %s)", statusCode, http.StatusText(statusCode))
+		return false, err
 	case statusCode == http.StatusNotFound: // don't retry on 404 not found
-		return false, fmt.Errorf("the resource you requested could not be found. (Error %d: %s)", statusCode, http.StatusText(statusCode))
+		return false, err
 	case statusCode >= 400 && statusCode < 500: // retry on 4xx client error
-		return true, fmt.Errorf("the server responded with an error. (Error %d: %s)", statusCode, http.StatusText(statusCode))
+		return true, err
```

Comment:
i think classically any 4xx error is a definite failure. my understanding was that the sdk never retries those when it doesn't know what they mean (like 429), only 5xx errors

Reply:
that's for retrying upsert run in python: https://github.com/wandb/wandb/blob/main/wandb/sdk/internal/internal_api.py#L2210-L2213

Reply:
https://github.com/wandb/wandb/blob/main/wandb/util.py#L881